Update plugin system to use registration

6utt3rfly opened the pull request:
(This PR is into the plugin-object PR instead of main)
In order to build a separate output dis file per plugin, I thought we should use a plugin registration system. It only needs a name and an `init` function so that plugin can configure jsep and add hooks as needed. If the plugin is registered a second time, the name will prevent the init function from being run again.

- Updated Object + Ternary plugins. Ternary plugin is registered by default
- Updated rollup to build each plugin as a separate output (all formats). Also delete dist folder before building.
- don't pollute the JSEP namespace from plugins. Use local constants instead.

6utt3rfly committed:
Update plugin system to support separate-files for each plugin with dependency injection
- Added plugin registration to JSEP. Plugins have a `name` and an `init` function that is called on registration with JSEP.
- Updated Object + Ternary plugins. Ternary plugin is registered by default
- Updated rollup to build each plugin as a separate output (all formats). Also delete dist folder before building.
- don't pollute the JSEP namespace from plugins. Use local constants instead.

LeaVerou commented on README.md:
```diff
@@ -87,12 +87,53 @@ jsep.removeIdentifierChar('@');
+
+#### Writing Your Own Plugin:
+Plugins are objects with two properties: `name` and `init`.
+Here's a simple plugin example:
```
Why is `init` needed? 
The plugins can just import `jsep` and do whatever to it.

LeaVerou commented on README.md:
````diff
@@ -87,12 +87,53 @@ jsep.removeIdentifierChar('@');
+```javascript
+import jsep from 'jsep';
+import object from 'jsep/plugins/object.js';
+jsep.plugins.register(object);
````
Why is `register()` needed? Importing the plugin should automatically register it, unless there's a reason not to.

LeaVerou commented on README.md:
```diff
@@ -87,12 +87,53 @@ jsep.removeIdentifierChar('@');
+Plugins are registered by calling `jsep.plugins.register(the_plugin)`.
+
+#### JSEP-provided plugins:
+* `ternary`: Built-in by default, adds support for ternary `a ? b : c` expressions
```
Should we add some wording to indicate that we plan to eventually remove this from core jsep and make it an optional plugin?

6utt3rfly replied:
Yeah probably worth mentioning ... depends on when/if we want a breaking change release?

LeaVerou commented on README.md:
```diff
@@ -87,12 +87,53 @@ jsep.removeIdentifierChar('@');
+##### Hooks
+Most plugins will make use of hooks to modify the parsing behavior of jsep.
+All hooks are bound to the jsep instance, are called with a single argument, and return void.
+The 'this' context provides access to the internal parsing methods of jsep
```
```suggestion
The `this` context provides access to the internal parsing methods of jsep
```

LeaVerou commented:
Ah, I just now re-read your post, sorry about the first two questions.
So the reason for the registration system is to prevent double registration?
If so, I'm not sure it's needed. Importing the same plugin from multiple places still only imports it once, the JS implementation takes care of that. Obviously this only happens when the plugin is imported from the same URL. But is the case of importing the same plugin multiple times using different URLs prevalent enough to design the entire API around it and introduce extra steps in every single plugin inclusion?

6utt3rfly commented:
> So the reason for the registration system is to prevent double registration?
Only the `name` is needed to prevent double-registration. It could also just check for the same function pointer though, too, if we really wanted?

However, the biggest reason for registration is more about building and deploying the plugin system. If importing the plugin needs to work without registration, then the plugin needs to be able to find the correct jsep module to attach itself to. Normally rollup would build the dependencies into the plugin, which means that a separate copy of the jsep code would be bundled with each plugin... which means that the plugin basically wouldn't work on the `import from 'jsep'` copy. We could also bundle all plugins with the main, jsep build output, but then the plugins would still need something to force its usage and what about 3rd-party plugins?

If we block rollup from including jsep in each plugin build, then it would need to leave the import statement. So then that import statement needs to find jsep properly in all cases. If a user uses, say, `dist/cjs/jsep.cjs.min.js`, then how does the plugin find jsep to attach itself to? `import ./jsep.cjs.min.js`? Using `meta.url` is not standardized, but it only helps with the path which isn't needed if using relative paths. In node CJS, it's possible to use `require.cache` but that doesn't exist in ESM. It's possible to manually replace the import string on every build output format, but I'm not sure that's the best way to do it, because what about 3rd-party plugins?

Maybe I'm missing an alternative solution though?

6utt3rfly commented:
@LeaVerou I was looking at rollup.js [plugins](https://rollupjs.org/guide/en/#a-simple-example) and they have a registration system (`plugins: [...]`) where each plugin is an object with various methods: `{name, load(), buildStart(), ...}`. The plugins don't register hook listeners, but they provide one by including the key in the plugin object. It's pretty similar to this PR, except that here it's just an init() method and each hook has to be registered by the plugin. It wouldn't be hard to change it to be more like rollup if wanted? Or do you have further thoughts on any way to have a plugin find and attach to Jsep without a registration method?

LeaVerou commented:
Hey, sorry for the delay in responding, I was on vacation, and then too busy.

After thinking about it for a while, I think it's not too big a compromise to use some sort of registration system, though I think it should be possible to register multiple plugins at once. It does put more effort on the user, but it's also more flexible. However, how does that work if `jsep` is used as an old-style IIFE in a browser? We don't want to prevent this use case from utilizing plugins, or restrict it to a specific bundle.

6utt3rfly committed:
Update README.md with backticks

Co-authored-by: Lea Verou <lea@verou.me>

6utt3rfly committed:
Allow registering multiple plugins at once
(single object arg, single array of objects arg, multiple objects arg)

6utt3rfly commented:
> However, how does that work if `jsep` is used as an old-style IIFE in a browser? We don't want to prevent this use case from utilizing plugins, or restrict it to a specific bundle.

I created this simple HTML file to test IIFE and it works as expected (but perhaps `object` should be renamed to `jsepObject` or `jsep-object-plugin` or something?
```html
<html>
  <head>
    <script src="../dist/iife/jsep.iife.min.js"></script>
    <script src="../dist/iife/object.iife.min.js"></script>
    <script>
      jsep.plugins.register(object);
      let result = jsep('({ a: 1, b: 2 })');
      console.log(JSON.stringify(result, null, 2));
    </script>
  </head>
</html>
```

6utt3rfly commented:
> I think it should be possible to register multiple plugins at once

I added support for registering multiple plugins in any preferred way:
```javascript
jsep.plugins.register(plugin)
jsep.plugins.register(plugin1, plugin2);
jsep.plugins.register([plugin1, plugin2]);
```

LeaVerou commented on README.md:
```diff
@@ -87,12 +87,58 @@ jsep.removeIdentifierChar('@');
 JSEP supports defining custom hooks for extending or modifying the expression parsing.
-All hooks are bound to the jsep instance and called with a single argument and return void.
-The 'this' context provide access to the internal parsing methods of jsep
+Plugins are registered by calling `jsep.plugins.register(the_plugin)`.
```
How do authors that wish to use plugins find what `the_plugin` identifier to use? To us it may be obvious, but to a beginner reading the docs, it is very opaque. 

Also, I think we should recommend the aggregate way of registering plugins, i.e. you import all the plugins, then you register them all.

LeaVerou commented on src/plugins.js:
```diff
@@ -0,0 +1,37 @@
+	 */
+	register(...plugins) {
+		plugins.forEach((plugin) => {
+			if (Array.isArray(plugin)) {
```
Do we really need arrays too? What's the use case?

LeaVerou commented:
> > However, how does that work if `jsep` is used as an old-style IIFE in a browser? We don't want to prevent this use case from utilizing plugins, or restrict it to a specific bundle.
> 
> I created this simple HTML file to test IIFE and it works as expected (but perhaps `object` should be renamed to `jsepObject` or `jsep-object-plugin` or something?
> 
> ```
> <html>
>   <head>
>     <script src="../dist/iife/jsep.iife.min.js"></script>
>     <script src="../dist/iife/object.iife.min.js"></script>
>     <script>
>       jsep.plugins.register(object);
>       let result = jsep('({ a: 1, b: 2 })');
>       console.log(JSON.stringify(result, null, 2));
>     </script>
>   </head>
> </html>
> ```

Yeah, we don't want to be creating globals that are named this generically! I'm not even super comfortable with creating a global for each plugin, that's too many globals even for IIFE. What can we do?

6utt3rfly commented:
> What can we do?

We could create a rollup file to bundle all plugins in one?

Or we could recommend registering plugins via esm over IIFE?
```html
<script src="../dist/iife/jsep.iife.min.js"></script>
<!--		<script src="../dist/iife/object.iife.min.js"></script>-->
<script type="module">
	import object from '../dist/object.min.js';
	jsep.plugins.register(object);
	let result = jsep('({ a: 1, b: 2 })');
	console.log(JSON.stringify(result, null, 2));
</script>
```

LeaVerou commented:
We definitely don't want to depend on bundling all plugins on a single file, that doesn't solve the problem of third-party plugins, or additional future plugins.

Recommending against IIFE also doesn't work, those who use IIFE today are primarily people who can't use ESM, most likely because they learned JS years ago and have not had time to catch up. A lot of academia falls in that bucket (where this project started, AFAIK, so we don't want to alienate that user base). 

Her's another possibly bad idea: what if plugins check for the presence of a global and auto-register themselves in that case? Then we don't need to create any global for them. 

This reminds me, is there a data structure where authors can inspect which plugins are already registered?

6utt3rfly commented:
> Here's another possibly bad idea: what if plugins check for the presence of a global and auto-register themselves in that case? Then we don't need to create any global for them.

It's possible with something like:
```
plugin = { name: 'plugin', init(jsep) {...} }

// if jsep is globally defined (IIFE), then auto-register:
if (typeof jsep === 'function') {
  plugin.init(jsep);
}
```
... its puts a little bit more work on other plugin authors ... and feels a little inconsistent for working differently in IIFE vs CJS/ESM, but is also a nice convenience... I don't think I feel too strongly either way 🤷🏻‍♀️ ??



> is there a data structure where authors can inspect which plugins are already registered?

the `plugins.js` file just uses a simple object `{ [key: name]: plugin }` which could be accessed through `jsep.plugins.plugins` 🙄 . I wasn't thinking much about a usage for reading the plugins (it already prevents duplicates), but I don't know? Thoughts?

LeaVerou commented:
Thinking about my last idea more, I think it's a terrible one. What if a library imports jsep, on a page that happens to already use the IIFE version? Suddenly the library's imports will start behaving weirdly. Also, I don't like requiring more boilerplate of plugin authors.

What about this lighter version instead? If there is a global, the plugin adds itself to an object of available plugins (e.g. `jsep.plugins.available`). Then, it can be registered, either by pointing to it (e.g. `jsep.plugins.available.object`) or by using its identifier (`jsep.plugins.register("object")`).

Although that would also require more boilerplate... Maybe we should just go with the global, and just make sure it's more specific, like e.g. `jsepObject`.

6utt3rfly committed:
Remove array arg support for plugin registration

6utt3rfly committed:
Rename plugins to prefix with `jsep`

6utt3rfly commented:
> Although that would also require more boilerplate... Maybe we should just go with the global, and just make sure it's more specific, like e.g. `jsepObject`.

It's unfortunate that automatic registration in all use-cases doesn't seem to be that possible... and I think making a `plugins.available` object would have the same issues as automatic registration in IIFE vs ESM/CJS (and would require somehow registering them at some point). I think renaming is probably our best option here, although automatic registration in any use case would have certainly been nice! I committed a change to prefix the plugins with `jsep`, so like `jsepObject` and `jsepTernary`. 

If you're good with this, I'm thinking we could treat the `plugin-object` branch (#144) as a feature branch to base other plugin PRs off of, then merge all to main when ready?

6utt3rfly committed:
- rename `jsep.plugins.plugins` => `jsep.plugins.registered` and add to readme
- gobbleExpression and gobbleToken need to stop running hooks once a node is found, similar to how jsep checks when calling each successive gobble method now. This saves each hook from having to check for `env.node` (and potentially leading to future errors with 2 hooks parsing 2 sequential expressions as just the last node). Added unit tests.

6utt3rfly committed:
When testing, cleanup hooks before/after to avoid cross-interference between plugins and tests

LeaVerou commented:
TBH I'm not super happy with that but it's more important that we move forward with the plugin system than to debate this. This PR has been open for 40 days at this point, let's merge it so we can start merging the plugin PRs too! I'll leave that to you, @6utt3rfly!
